Mathlib/RingTheory/TwoSidedIdeal/Commute.lean

```diff
@@ -0,0 +1,38 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir, María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+import Mathlib.Data.Nat.Choose.Sum
+import Mathlib.RingTheory.TwoSidedIdeal.BigOperators
+
+/-! # Powers of commuting elements in two sided ideals -/
+
+namespace TwoSidedIdeal
+
+variable {R : Type*} [Ring R] (I : TwoSidedIdeal R) {a b : R}
+
+theorem pow_mem_of_pow_mem {m n : ℕ} (ha : a ^ m ∈ I) (h : m ≤ n) : a ^ n ∈ I := by
+  rw [← Nat.add_sub_of_le h, pow_add]
+  exact I.mul_mem_right _ _ ha
+
+theorem add_pow_mem_of_pow_mem_of_le {m n k : ℕ}
+    (ha : a ^ m ∈ I) (hb : b ^ n ∈ I) (hk : m + n ≤ k + 1) (h : Commute a b) :
+    (a + b) ^ k ∈ I := by
+  rw [h.add_pow]
+  apply I.finsetSum_mem
+  intro c _
+  apply mul_mem_right
+  by_cases h : m ≤ c
+  · exact I.mul_mem_right _ _ (I.pow_mem_of_pow_mem ha h)
+  · refine I.mul_mem_left _ _ (I.pow_mem_of_pow_mem hb ?_)
+    omega
+
+theorem add_pow_add_pred_mem_of_pow_mem {m n : ℕ}
+    (ha : a ^ m ∈ I) (hb : b ^ n ∈ I) (h : Commute a b) :
+    (a + b) ^ (m + n - 1) ∈ I :=
+  I.add_pow_mem_of_pow_mem_of_le ha hb (by rw [← Nat.sub_le_iff_le_add]) h
```

Comment:
It sounds to me that all of these should be true for left ideals as well thanks to the commutation hypotheses, right? I recognize it's a bit unsatisfying to be able to state them for left ideals but not right ideals, but I think it would still be better not to restrict to two-sided ideals.

Reply:
Yeah, it seems you're right.

Reply:
On the other hand, having this for ideals requires to generalize (it is often trivial) lemmas from `CommSemiring` to `Semiring` in much earlier files of the library. So I believe I'll delete the noncommutative case.

Reply:
That shouldn't require that much generalizing, right? To be clear I think that would fit in this PR.